File: 2022/day07/php/solution.php
<?php

/**
 * Advent of Code 2022 - Day 7: No Space Left On Device
 *
 * Parse terminal output to build directory tree, track sizes,
 * and find directories meeting size criteria.
 */

function parseFilesystem(array $lines): array {
    $path = [];
    $dirSizes = [];

    foreach ($lines as $line) {
        if (str_starts_with($line, '$ cd')) {
            $target = substr($line, 5);
            if ($target === '/') {
                $path = ['/'];
            } elseif ($target === '..') {
                array_pop($path);
            } else {
                $path[] = $target;
            }
        } elseif (str_starts_with($line, '$ ls')) {
            continue;
        } elseif (str_starts_with($line, 'dir ')) {
            continue;
        } else {
            // It's a file with size
            $parts = explode(' ', $line);
            $size = (int)$parts[0];

            // Add size to current directory and all parent directories
            for ($i = 0; $i < count($path); $i++) {
                $dirPath = implode('/', array_slice($path, 0, $i + 1)) ?: '/';
                if (!isset($dirSizes[$dirPath])) {
                    $dirSizes[$dirPath] = 0;
                }
                $dirSizes[$dirPath] += $size;
            }
        }
    }

    return $dirSizes;
}

function part1(array $dirSizes): int {
    $sum = 0;
    foreach ($dirSizes as $size) {
        if ($size <= 100000) {
            $sum += $size;
        }
    }
    return $sum;
}

function part2(array $dirSizes): int {
    $totalSpace = 70000000;
    $neededSpace = 30000000;
    $usedSpace = $dirSizes['/'];
    $freeSpace = $totalSpace - $usedSpace;
    $needToFree = $neededSpace - $freeSpace;

    // Find smallest directory >= needToFree
    $candidates = array_filter($dirSizes, fn($size) => $size >= $needToFree);
    return min($candidates);
}

function main(): void {
    $inputFile = __DIR__ . '/../input.txt';
    $content = trim(file_get_contents($inputFile));
    $lines = explode("\n", $content);

    $dirSizes = parseFilesystem($lines);

    echo 'Part 1: ' . part1($dirSizes) . PHP_EOL;
    echo 'Part 2: ' . part2($dirSizes) . PHP_EOL;
}

main();

File: 2022/day07/clojure/solution.clj
#!/usr/bin/env clojure -M

(require '[clojure.string :as str])

(defn parse-filesystem
  "Parse terminal output and return map of directory path to total size."
  [lines]
  (loop [lines lines
         path []
         dir-sizes {}]
    (if (empty? lines)
      dir-sizes
      (let [line (first lines)
            rest-lines (rest lines)]
        (cond
          ;; cd command
          (str/starts-with? line "$ cd")
          (let [target (subs line 5)]
            (cond
              (= target "/")
              (recur rest-lines ["/"] dir-sizes)

              (= target "..")
              (recur rest-lines (vec (butlast path)) dir-sizes)

              :else
              (recur rest-lines (conj path target) dir-sizes)))

          ;; ls command - skip
          (str/starts-with? line "$ ls")
          (recur rest-lines path dir-sizes)

          ;; dir entry - skip
          (str/starts-with? line "dir ")
          (recur rest-lines path dir-sizes)

          ;; file entry - add size to all parent directories
          :else
          (let [size (Long/parseLong (first (str/split line #" ")))
                ;; Build all path prefixes and add size to each
                updated-sizes (reduce
                               (fn [sizes i]
                                 (let [dir-path (str/join "/" (take (inc i) path))]
                                   (update sizes dir-path (fnil + 0) size)))
                               dir-sizes
                               (range (count path)))]
            (recur rest-lines path updated-sizes)))))))

(defn part1
  "Sum of sizes of directories with total size <= 100000."
  [dir-sizes]
  (->> (vals dir-sizes)
       (filter #(<= % 100000))
       (reduce +)))

(defn part2
  "Find smallest directory to delete to free enough space."
  [dir-sizes]
  (let [total-space 70000000
        needed-space 30000000
        used-space (get dir-sizes "/")
        free-space (- total-space used-space)
        need-to-free (- needed-space free-space)]
    (->> (vals dir-sizes)
         (filter #(>= % need-to-free))
         (apply min))))

(defn -main []
  (let [script-dir (.getParent (java.io.File. *file*))
        input-file (str script-dir "/../input.txt")
        lines (-> (slurp input-file)
                  str/trim
                  (str/split #"\n"))
        dir-sizes (parse-filesystem lines)]
    (println "Part 1:" (part1 dir-sizes))
    (println "Part 2:" (part2 dir-sizes))))

(-main)
